### PR TITLE
Add capture panic functionality

### LINKED ISSUE #35: Handle  GoPanics as errors
### Context

As a software developer
I want unexpected panics on my program to be supervised
So that my system becomes more resilient

### Acceptance Criteria

* [ ] Panics are captured by default
* [ ] We have an option that allows us to not capture panics
* [ ] Add test cases that assert functionality is working

### PR DESCRIPTION
**Pull request by roman**
Closes #35 

### Context

As a software developer
I want unexpected panics on my program to be supervised
So that my system becomes more resilient

### Acceptance Criteria

* [x] Panics are captured by default
* [x] We have an option that allows us to not capture panics
* [x] Add test cases that assert functionality is working

### CODE REVIEW

File: internal/c/start.go [hunk 1]
```diff
@@ -58,11 +95,11 @@ func waitTimeout(
 // logic.
 //
 func (chSpec ChildSpec) DoStart(
-	parentName string,
+	supName string,
```
Comment (julio-garcia-fc) [inline internal/c/start.go:98]: Much better names. Thank you.

File: internal/c/start.go [hunk 2]
```diff
@@ -78,6 +115,29 @@ func (chSpec ChildSpec) DoStart(
 		// we cancel the childCtx on regular termination
 		defer cancelFn()
 
+		defer func() {
```
Comment (julio-garcia-fc) [inline internal/c/start.go:118]: Do you need to remove the `TODO:` above?

File: internal/c/start.go [hunk 3]
```diff
@@ -41,6 +42,42 @@ func waitTimeout(
 	}
 }
 
+func sendChildNotification(
```
Comment (julio-garcia-fc) [inline internal/c/start.go:45]: The name threw me off. It is a little ambiguous. Would you consider `sendNotificationToSupervisor()`?

File: internal/stest/workers.go [hunk 4]
```diff
@@ -95,8 +95,43 @@ func FailOnSignalWorker(
...
+	currentFailCount := int32(0)
+	startCh := make(chan struct{})
+	startSignal := func(done bool) {
+		if done {
+			close(startCh)
+			return
+		}
+		startCh <- struct{}{}
+	}
+	opts = append(opts, cap.WithCapturePanic(true))
```
Comment (julio-garcia-fc) [inline internal/stest/workers.go:116]: It's the default anyway, right?

File: cap/capture_panic_test.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		)
+
+		assert.NoError(t, err)
+
+		AssertExactMatch(t, events,
+			[]EventP{
+				// start children from left to right
+				WorkerStarted("root/child1"),
+				SupervisorStarted("root"),
+				WorkerFailedWith("root/child1", "Panicking child (1 out of 1)"),
```
Comment (julio-garcia-fc) [inline cap/capture_panic_test.go:40]: This is counter intuitive. The progression that makes sense to me is one of the following:

1. Worker attempts to start
2. panics while starting
3. fails to start
4. terminates

OR

1. worker attempts to start
2. worker successfully starts
3. while running business logic, it panics
4. terminates

The sequence you describe, with a failure, an apparent successful start, followed by an abrupt terminate, looks strange. Seems like you need a `WorkerFailedToStart()` predicate? Is that what you mean?

Comment (roman) [reply cap/capture_panic_test.go:40]: I'll add comments explaining the sequence

File: internal/stest/workers.go (hunk 4 above)
Comment (roman) [reply internal/stest/workers.go:116]: Yup, just making it explicit that's all

File: internal/c/start.go (hunk 3 above)
Comment (roman) [reply internal/c/start.go]: Good idea!

File: internal/c/start.go (hunk 2 above)
Comment (roman) [reply internal/c/start.go:118]: Good catch

File: cap/capture_panic_test.go [hunk 6]
```diff
@@ -34,11 +35,23 @@ func TestCapturePanic(t *testing.T) {
...
+
 				WorkerStarted("root/child1"),
 				SupervisorStarted("root"),
+				// ^^^ supervisor started successfully, from here, it is supervising the
+				// only worker node it has
 				WorkerFailedWith("root/child1", "Panicking child (1 out of 1)"),
+				// ^^^ As specified with the PanicOnSignalWorker, this worker fails
+				// only one time (1 out of 1)
 				WorkerStarted("root/child1"),
+				// ^^^ Worker is (re)started, this time, it won't panic (given we ran)
```
Comment (julio-garcia-fc) [inline cap/capture_panic_test.go:48]: Got it!!! Thank you! I missed the implied restart policy. The policies are awesome, but hard to keep straight.

**Pull request merged**